tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -462,32 +444,107 @@ add_data_node_scan_paths(PlannerInfo *root, RelOptInfo *baserel)
 		}
 	}
 
+	return ppi_list;
+}
+
+static void
+add_data_node_scan_paths(PlannerInfo *root, RelOptInfo *data_node_rel, RelOptInfo *hyper_rel,
+						 List *ppi_list)
+{
+	TsFdwRelInfo *fpinfo = fdw_relinfo_get(data_node_rel);
+	Path *path;
+
+	if (data_node_rel->reloptkind == RELOPT_JOINREL)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("foreign joins are not supported")));
+
+	path = data_node_scan_path_create(root,
+									  data_node_rel,
+									  NULL, /* default pathtarget */
+									  fpinfo->rows,
+									  fpinfo->startup_cost,
+									  fpinfo->total_cost,
+									  NIL, /* no pathkeys */
+									  NULL,
+									  NULL /* no extra plan */,
+									  NIL);
+
+	fdw_utils_add_path(data_node_rel, path);
+
+	/* Add paths with pathkeys */
+	fdw_add_paths_with_pathkeys_for_rel(root, data_node_rel, NULL, data_node_scan_path_create);
+
 	/*
-	 * Now build a path for each useful outer relation.
+	 * Now build a path for each useful outer relation, if the parameterized
+	 * data node scans are not disabled.
 	 */
-	foreach (lc, ppi_list)
+	if (!ts_guc_enable_parameterized_data_node_scan)
+	{
+		return;
+	}
+
+	ListCell *ppi_cell;
+	foreach (ppi_cell, ppi_list)
 	{
-		ParamPathInfo *param_info = (ParamPathInfo *) lfirst(lc);
+		ParamPathInfo *param_info = (ParamPathInfo *) lfirst(ppi_cell);
+
+		/*
+		 * Guess if this data node scan is likely to be an index scan, and
+		 * factor that into the cost.
+		 */
+		double scan_cost_discount_multiplier = 1.;
+		ListCell *path_cell;
+		foreach (path_cell, hyper_rel->pathlist)
+		{
+			Path *path = (Path *) lfirst(path_cell);
+			if (path->param_info == param_info)
+			{
+				/*
+				 * We shouldn't have parameterized seq scans. Can be an
+				 * IndexPath (includes index-only scans) or a BitmapHeapPath.
+				 */
+				Assert(path->type == T_BitmapHeapPath || path->type == T_IndexPath);
+
+				/*
+				 * We have an index path with a matching parameterization for
+				 * the hyperatble rel. This means that we also have a matching
+				 * index on the data nodes. As a hack, try to cost this path not
+				 * as a seq scan, but as an imaginary index scan that's cheaper
+				 * by a constant factor.
+				 */
+				scan_cost_discount_multiplier = 0.1;
+				break;
+			}
+		}
+
+		/*
+		 * As a baseline, cost the data node scan as a seq scan.
+		 */
 		Cost startup_cost = 0;
 		Cost run_cost = 0;
-		double rows = baserel->tuples > 1 ? baserel->tuples : 123456;
+		double rows = data_node_rel->tuples > 1 ? data_node_rel->tuples : 123456;
```

Comment:
hmm that is weird i think postgres uses a different convention for unknown tuple estimates any reason why we are not following that

Reply:
You mean what `table_block_relaion_estimate_size` does when there are no stats? It knows the number of pages, but we don't, since this is a distributed table. The particular number is arbitrary, I just chose something that would be easy to spot when looking at EXPLAIN.